src/ray/ray_syncer/ray_syncer_bidi_reactor_base.h

```diff
@@ -119,36 +153,49 @@ class RaySyncerBidiReactorBase : public RaySyncerBidiReactor, public T {
   }
 
   void StartSend() {
-    if (sending_) {
+    if (sending_ || sending_buffer_.empty()) {
       return;
     }
 
-    if (sending_buffer_.size() != 0) {
-      auto iter = sending_buffer_.begin();
-      auto msg = std::move(iter->second);
-      sending_buffer_.erase(iter);
-      Send(std::move(msg), sending_buffer_.empty());
-      sending_ = true;
+    // Cancel any pending batch timer since we're sending now
+    if (batch_timer_active_) {
+      batch_timer_.cancel();
+      batch_timer_active_ = false;
```

Comment:
Hm I'm not sure this is correct. I think we should only be resetting the timer from within `PushToSendingQueue`. `StartSend` is called from within `SendNext`, which is triggered from `OnWriteComplete`.

This can cause the following order of events:

- Send batch from `OnWriteComplete`
- New message comes in to `PushToSendingQueue`, start timer
- `OnWriteComplete` fires from first batch, which cancels the timer
- First batch finishes sending, no new messages are pushed to sending queue for a long time, the message that was pushed to start the timer hangs in the queue indefinitely